---
title: "AA-GP: Figures"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup}
knitr::opts_knit$set(root.dir = "~/Documents/AA-GenomicPrediction")

library(tidyverse)
library(broom)
library(ggrepel)

# set colorblind friendly palette
cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")
```

```{r load results}
load("reports/gblup.RData")
load("reports/model_summary.RData")
load("~/Documents/AA-GenomicPrediction/reports/model_summary.RData")

```

## Figure 1 - GBLUP

```{r fig 1}
gblup_summary %>%
  mutate(aa_cat = fct_relevel(aa_cat, "absolute", "relative", "family"),
         family = fct_relevel(family, "aromatic", after = 4)) %>%
  group_by(trait) %>%
  filter(mean(gblup_pa) > 0.3) %>%
  ungroup() %>%
  ggplot(., aes(reorder(trait, as.numeric(as.factor(aa_cat))), gblup_pa, fill = aa_cat, col = aa_cat)) +
  scale_fill_manual(name = "", values = c("#66a61e", "#d95f02", "#7570b3")) + 
  scale_colour_manual(name = "", values = c("#66a61e", "#d95f02", "#7570b3")) + 
  geom_point(position = position_jitterdodge(dodge.width = 1), size = 0.75) + 
  geom_boxplot(col = "grey25", position = position_dodge(width = 0.9), 
               outlier.colour = NA, width = 0.5, alpha = 0.5) + 
  geom_point(aes(trait, gblup_h2), colour = "black", shape = 2, size = 1.25,
             show.legend = FALSE) + 
  facet_grid(.~ family, scales = "free_x", space = "free") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 10)) +
  xlab("") +
  ylab("Prediction accuracy (r)")


ggsave("reports/figures/Fig1.pdf", height = 3, width = 7.5)

```

## Figure 2 - MultiBLUP results

```{r multiblup results, results = "hide"}
fam <- data.frame(trait = gblup_summary$trait, family = gblup_summary$family,
                  aa_cat = gblup_summary$aa_cat)

multiblup_pass <- read_csv("reports/multiblup_pass.csv") %>%
  complete(pathway, nesting(trait)) %>%
  left_join(., fam) %>%
  mutate(family = factor(family, 
                         levels = c("aspartate", "glutamate", "BCAA_pyruvate", 
                                    "serine", "aromatic", "total")),
         trait = fct_reorder(trait, as.numeric(family)),
         aa_cat = fct_relevel(aa_cat, 
                              "relative", 
                              after = 1),
         aa_cat = fct_recode(aa_cat, 
                             abs = "absolute", 
                             rel = "relative", 
                             fam = "family"),
         family = fct_recode(family, 
                             "pyruvate (BCAA)" = "BCAA_pyruvate"),
         category = ifelse(
      grepl("degradation_|protein_|protease", pathway),
      "protein",
      ifelse(
        grepl(
          "flavonoids|isoprenoids|N_containing|phenylpropanoids|S_containing",
          pathway
        ),
        "specialized",
        ifelse(
          grepl("aa_degradation|aa_synthesis|aa_transport", pathway),
          "amino\nacid",
          "primary"
        )
      )
    ) ,
    category = factor(
      category,
      levels = c("amino\nacid", "primary", "specialized", "protein")
    )) %>% 
  dplyr::select(trait, pathway, multiblup_h2, family, aa_cat, category) %>%
  distinct() 
```

```{r fig 2}
multiblup_pass %>%
    ggplot(aes(trait, fct_rev(pathway), 
               size = multiblup_h2, 
               colour = family)) +
    geom_point() +
    scale_colour_manual(name = "Family", 
                        values = cbPalette[c(5, 6, 7, 1, 3, 2)], 
                        drop = FALSE) +
    scale_size_continuous(range = c(0, 4),
                          limits = c(0, 1),
                          breaks = c(0.25, 0.5, 0.75, 1),
                          name = "Proportion of heritability explained") +
    facet_grid(category ~ family + aa_cat,
               scales = "free",
               space = "free",
               switch = "y") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          panel.spacing.y = unit(0.1, "lines"),
          panel.spacing.x = unit(0.1, "lines"),
          panel.grid.major = element_line(colour = 'lightgray', size = 0.25),
          text = element_text(size = 10),
          axis.text = element_text(size = 8),
          strip.text = element_text(size = 8, margin = margin()),
          strip.text.y = element_text(angle = 180),
          legend.position = "top") +
    labs(y = "Pathway", x = "Trait") +
    guides(colour = FALSE, 
           size = guide_legend(ncol = 4, byrow = TRUE)) +
    scale_y_discrete(position = "right") 

# ggsave("reports/figures/Fig2.eps", height = 4, width = 7.5, device = "eps")

```

## Figure 3 - GBLUP vs MultiBLUP

```{r read in data}
# multiblup results
multiblup_pass <- read_csv("reports/multiblup_pass.csv")

# gwas results
load("data/raw/gwas_results.RData")

exclude <- c("aa_deg_bcat", "aa_syn_nobcat","aa_synthesis", "aa_degradation",
             "asp_protease", "cys_protease", "degradation_AAA", "degradation_autophagy",
             "degradation_subtilases", "glycolysis_cystolic", "glycolysis_plastid", "glycolysis_other",
             "metallo_protease", "protein_degradation_old", "ser_protease")
  
pathways <- tibble(file = list.files(path="data/processed/pathways", 
                                     pattern = "*.txt", 
                                     full.names = TRUE, 
                                     recursive = TRUE)) %>%
  separate(file, sep = "/", 
           into = c("dir", "dir2", "dir3", "pathway"),
           remove = FALSE) %>%
  mutate(data = lapply(file, read.table, header = TRUE, na = c("", "NA")),
         data = lapply(data, mutate_all, as.character)) %>%
  unnest(data) %>%
  mutate(pathway = gsub(".txt", "", pathway)) %>%
  filter(!pathway %in% exclude) %>%
  mutate(position = as.numeric(position),
         pathway = fct_recode(pathway,
                              aa_degradation = "aa_deg_bcathaplo",
                              aa_synthesis = "aa_syn_nobcathaplo")) 

# combine with multiblup results
pathways <- left_join(pathways, multiblup_pass) %>%
  select(trait, pathway, chromosome, snp_id, position, 
         ensembl_gene_id.y, external_gene_name, BINCODE, NAME)
```


```{r duplicated SNPs}
# find duplicate snps
pathways %>%
  filter(snp_id %in% unique(.[["snp_id"]][duplicated(.[['snp_id']])])) %>%
  group_by(pathway) %>%
  tally()
```

# adjust FDR by pathway
```{r}
joined <- full_join(gwas_results, pathways, c("Position" = "position",
                                              "trait")) 

fdr_corrected <- joined %>% 
  group_by(trait, pathway) %>%
  drop_na(pathway, Chromosome) %>%
  mutate(new_fdr = p.adjust(P.value, 
                            method = "fdr", 
                            n = length(snp_id)),
         fdr_n = length(snp_id)) %>%
  ungroup() 

```

```{r}
# manhattan plot style from https://www.r-graph-gallery.com/wp-content/uploads/2018/02/Manhattan_plot_in_R.html
plotdf <- joined %>%
  # chr length
  group_by(Chromosome) %>%
  summarize(chr_len = max(Position)) %>%
  # cumulative position for each chr
  mutate(tot = cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # join with data
  left_join(joined, ., by = c("Chromosome")) %>%
  full_join(fdr_corrected, .) %>% 
  # add cumulative snp position
  arrange(chromosome.y, Position) %>%
  mutate(BP = Position + tot) %>%
  # highlight info
  mutate(highlight = ifelse(new_fdr < 0.10, "yes", "no"),
         external_gene_name = 
             coalesce(external_gene_name, ensembl_gene_id.y)) %>%
  filter(!is.na(Chromosome))

# x axis 
axisdf <- plotdf %>% 
  group_by(Chromosome) %>% 
  summarize(center = (max(BP) + min(BP))/2)

sig_traits <- plotdf %>%
  filter(new_fdr <= 0.10) %>%
  select(pathway) %>%
  distinct() %>%
  as_tibble()

```

```{r table 4}
aa_fam <- unique((plotdf$trait)) %>%
  as_tibble() %>%
  mutate(
         # add amino acid biochemical family ids (based on precursor)
         family = ifelse(grepl("asp|thr|met|asp|AspFam", value) & !grepl("ile", value), 
                         "aspartate", 
                         ifelse(grepl("arg|glu|gln|pro|his|GluFam", value),
                                "glutamate",
                                ifelse(grepl("gly|ser|SerFam", value),
                                       "serine",
                                       ifelse(grepl("phe|tyr|trp|ShikFam", value),
                                              "aromatic",
                                              ifelse(grepl("Total", value),
                                                     "total", "BCAA_pyruvate"))))),
         # add type of measurement (absolute, relative, ratio)
         aa_cat = ifelse(!grepl("_", value), "absolute",
                         ifelse(grepl("_t", value), "relative",
                                "family"))) %>%
  rename(trait = "value")

head(aa_fam)
```

```{r}
s3table <- plotdf %>%
  left_join(aa_fam, plotdf, by = "trait") %>%
  filter(new_fdr <= 0.10) %>%
  select(family, trait, pathway, Chromosome, Position, P.value, fdr_n, 
         `FDR_Adjusted_P-values`, ensembl_gene_id.y, external_gene_name) %>%
  arrange(family, trait, pathway) %>%
  write_csv("reports/s3table.csv")

head(s3table)
```

```{r}
p <- list() 

for (i in sig_traits$pathway) {
  t <- paste(i)
  
  print(t)
  
  df <- plotdf %>%
    drop_na(pathway) %>% 
    filter(pathway %in% t) %>%
    mutate(external_gene_name = 
             coalesce(external_gene_name, ensembl_gene_id.y))
  
  p[[i]] <- ggplot(df, aes(x = BP, y = -log10(P.value))) +
      geom_point(aes(color = as.factor(Chromosome)), alpha = 0.8) +
      scale_colour_manual(values = rep(c("grey", "skyblue"), 5)) +
      scale_x_continuous(label = axisdf$Chromosome,
                         breaks = axisdf$center) +
      scale_y_continuous(expand = c(0,0.1)) +
      geom_point(data = subset(df, highlight == "yes"),
                 colour = "orange", size = 2) +
      geom_label_repel(data = subset(df, highlight == "yes"),
               aes(label = external_gene_name), size = 2) +
      theme_bw() +
      theme(legend.position = "none") +
      facet_wrap(~trait) +
      ggtitle(t) 
  print(p[[i]])
}
```


```{r}
his <- plotdf %>%
  filter(trait %in% "his") %>% 
  ggplot(., aes(x = BP, y = -log10(P.value))) +
  geom_point(aes(color = as.factor(Chromosome)), alpha = 0.3, size = 0.75) +
  scale_colour_manual(values = rep(c("grey90", "grey80"), 5)) +
  scale_x_continuous(label = axisdf$Chromosome,
                     breaks = axisdf$center) +
  scale_y_continuous(expand = c(0,0.1)) +
  geom_point(data = subset(plotdf, pathway == "protein_folding" &
                             trait == "his"),
             colour = "#5ab4ac", shape = 17) + 
  geom_point(data = subset(plotdf, highlight == "yes" &
                             pathway == "protein_folding" & 
                             trait == "his"),
             colour = "orange", size = 2, shape = 17) +
  geom_label_repel(data = subset(plotdf, highlight == "yes" &
                                   pathway == "protein_folding" &
                                   trait == "his"),
                   aes(label = external_gene_name), size = 1.5) +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0, size = 12)) +
  xlab("Chromosome") + 
  labs(subtitle = "His - Protein folding")


his_glufam <- plotdf %>%  
  filter(trait %in% "his_GluFam") %>% 
  ggplot(., aes(x = BP, y = -log10(P.value))) +
  geom_point(aes(color = as.factor(Chromosome)), alpha = 0.3, size = 0.75) +
  scale_colour_manual(values = rep(c("grey90", "grey80"), 5)) +
  scale_x_continuous(label = axisdf$Chromosome,
                     breaks = axisdf$center) +
  scale_y_continuous(expand = c(0,0.1)) +
  geom_point(data = subset(plotdf, pathway == "aa_degradation" &
                             trait == "his_GluFam"),
             colour = "#5ab4ac", shape = 17) + 
  geom_point(data = subset(plotdf, highlight == "yes" &
                             pathway == "aa_degradation" & 
                             trait == "his_GluFam"),
             colour = "orange", size = 2, shape = 17) +
  geom_label_repel(data = subset(plotdf, highlight == "yes" &
                                   pathway == "aa_degradation" &
                                   trait == "his_GluFam"),
                   aes(label = external_gene_name), size = 1.5) +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0, size = 12)) +
  xlab("Chromosome") + 
  labs(subtitle = "His_GluFam - Amino acid degradation")
```  

```{r}

multiblup_table <- 
  left_join(gblup_summary, multiblup_summary, by = c("trait", "cvnum", "fold")) %>%
  group_by(trait, pathway, cvnum, fold) %>% 
  group_by(trait, pathway) %>% 
  select(trait, pathway, gblup_pa, multiblup_pa, 
         gblup_reliability, multiblup_reliability,
         gblup_bias_int, multiblup_bias_int,
         gblup_bias_slope, multiblup_bias_slope,
         gblup_mse, multiblup_mse) 

lines <- data.frame(metric = c("accuracy", "reliability", "intercept", "slope", "MSE"),
                    intercept = c(0, 0, 0, 1, 0))

boxes1 <- multiblup_table %>%
  ungroup() %>% 
  # select(trait, pathway, pa_diff) %>%
  filter(pathway == "protein_folding" & trait == "his") %>% #|
         # pathway == "aa_degradation" & trait == "his_GluFam") %>%
  gather(key = "metric", value = "value", -trait, -pathway) %>%
  separate(metric, sep = "_", into = c("method", "metric"), extra = "merge") %>% 
  mutate(metric = fct_recode(metric, "accuracy" = "pa", "intercept" = "bias_int", "slope" = "bias_slope", "MSE" = "mse"),
         metric = fct_relevel(metric,
                              "accuracy", "reliability", "intercept", "slope", "MSE")) %>%
  ggplot(., aes(method, value, fill = method, col = method)) +
  scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + 
  scale_colour_manual(values = c("#d8b365", "#5ab4ac")) + 
  geom_hline(data = lines, aes(yintercept = intercept), col = "darkgray") + 
  geom_point(position = position_jitterdodge(dodge.width = .8, 
                                             jitter.width = 0.2), size = 0.5) + 
  geom_boxplot(col = "grey25", position = position_dodge(width = 0.8), 
               outlier.colour = NA, width = 0.5, alpha = 0.5) + 
  facet_wrap(. ~ metric, scales = "free", nrow = 1) +
  theme_bw() +
  theme(# axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1),
        legend.position = "none",
        text = element_text(size = 10),
        axis.title.y = element_blank()) +
  labs(subtitle = "His - Protein folding")

boxes1

boxes2 <- multiblup_table %>%
  ungroup() %>% 
  # select(trait, pathway, pa_diff) %>%
  filter(pathway == "aa_degradation" & trait == "his_GluFam") %>%
  gather(key = "metric", value = "value", -trait, -pathway) %>%
  separate(metric, sep = "_", into = c("method", "metric"), extra = "merge") %>% 
  mutate(metric = fct_recode(metric, "accuracy" = "pa", "intercept" = "bias_int", "slope" = "bias_slope", "MSE" = "mse"),
         metric = fct_relevel(metric,
                              "accuracy", "reliability", "intercept", "slope", "MSE")) %>%
  ggplot(., aes(method, value, fill = method, col = method)) +
  scale_fill_manual(values = c("#d8b365", "#5ab4ac")) + 
  scale_colour_manual(values = c("#d8b365", "#5ab4ac")) + 
  geom_hline(data = lines, aes(yintercept = intercept), col = "darkgray") + 
  geom_point(position = position_jitterdodge(dodge.width = .8, 
                                             jitter.width = 0.2), size = 0.5) + 
  geom_boxplot(col = "grey25", position = position_dodge(width = 0.8), 
               outlier.colour = NA, width = 0.5, alpha = 0.5) + 
  facet_wrap(. ~ metric, scales = "free", nrow = 1) +
  theme_bw() +
  theme(# axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        legend.position = "none",
        text = element_text(size = 10),
        axis.title.y = element_blank()) +
  labs(subtitle = "His_GluFam - Amino acid degradation") 

boxes2

```

```{r}
theme_set(theme_cowplot(font_size = 10))

pt1 <- plot_grid(boxes1, boxes2, nrow = 2, labels = c('A', 'B'))

pt2 <- plot_grid(his, his_glufam, ncol = 2, labels = c('C', 'D'))

plot_grid(pt1, pt2, nrow = 2, labels = NULL)

ggsave("reports/figures/Fig3.pdf", width = 7, height = 7, units = "in")
```


  