---
title: "AA-GP: process null distribution"
output: 
  html_notebook:
    toc: yes
    toc_float: yes
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documents/aagp_ldak")

library(tidyverse)
library(kableExtra)
library(quantreg)
library(data.table)
library(viridis)
library(emdbook)
library(gridExtra)

date <- format(Sys.time(), '%Y%m%d')

```

# GBLUP log likelihood 

```{r gblup llik}

## read in amino acid names + numeric codes 
aa_names <- data.frame(trait = colnames(read.table("data/processed/pheno_file", header = TRUE)[3:56]), id = as.character(1:54))

# group into categories by type of measurement (absolute, relative, ratio)
aa_cat <- list(absolute = data.frame(trait = aa_names[!grepl("_t|Fam|Corr|BCAA", aa_names$trait), 1], aa_cat = "absolute"),
               relative = data.frame(trait = aa_names[grepl("_t", aa_names$trait), 1], aa_cat = "relative"),
               family = data.frame(trait = aa_names[grepl("Fam|Corr|BCAA", aa_names$trait),1], aa_cat = "family")) %>%
  map_df(I)

## group into amino acid families (based on precursors)
# aa_fam <- list(
#   aspartate = data.frame(trait =
#                            aa_names[grepl("asp|ile|lys|thr|met|Asp", aa_names$trait) &
#                                       !grepl("GluFamCorr", aa_names$trait), 1], aa_fam = "aspartate"),
#   glutamate = data.frame(trait = aa_names[grepl("arg|glu|gln|pro|GluFam", aa_names$trait) &
#                                             !grepl("his", aa_names$trait), 1], aa_fam = "glutamate"),
#   serine = data.frame(trait = aa_names[grepl("gly|ser|SerFam", aa_names$trait), 1], aa_fam = "serine"),
#   pyruvate = data.frame(trait = aa_names[grepl("ala|val|leu|BCAA|PyrFam", aa_names$trait) &
#                                            !grepl("GluFamCorr|ile", aa_names$trait), 1], aa_fam = "pyruvate"),
#   shikimate = data.frame(trait = aa_names[grepl("phe|tyr|trp|ShikFam", aa_names$trait), 1], aa_fam = "shikimate"),
#   histidine = data.frame(trait = aa_names[grepl("his", aa_names$trait), 1], aa_fam = "histidine"),
#   total = data.frame(trait = "Total", aa_fam = "total")
# ) %>%
#   map_df(I)

# gblup_llik <- tibble(file = list.files(path = "models/gblup_h2",
#                                      pattern = "gblup_\\d+.reml$",
#                                      full.names = TRUE)) %>%
#   separate(file, sep = "/", into = c("source", "method", "id"), remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE)) %>%
#   unnest(data) %>%
#   filter(Num_Kinships %in% "Alt_Likelihood") %>%
#   mutate(id = str_replace(id, "gblup_", ""),
#          id = str_replace(id, ".reml", ""),
#          gblup_llik = as.numeric(X1)) %>%
#   left_join(., aa_names, by = "id") %>%
#   left_join(., aa_cat, by = "trait") %>%
#   select(id, trait, gblup_llik) %>%
#   write_csv(paste0("reports/", date, "_gblup_llik.csv"))

gblup_llik <- read_csv("reports/20190212_gblup_llik.csv") %>%
  mutate_at(vars(id), as.character) %>%
  left_join(., aa_cat, by = "trait")

```

# LR for null distribution 

```{r null summary}

## get size of each null category
## ran on cluster using summarize_random_subsets.R

# bcat_genes <- read.table("data/processed/pathways/bcat_genes/bcat_genes.txt", header=TRUE)
# 
# files <- list.files(path = "data/processed/random_sets", pattern = "*.txt", full.names = TRUE)
# 
# feature_sizes <- tibble(file = files) %>%
#   separate(file, sep = "/", into = c("dir", "dir2", "dir3", "pathway"), remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE)) %>%
#   unnest(data) %>%
#   mutate(pathway = str_replace(pathway, "null", "c"),
#          pathway = str_replace(pathway, ".txt", "")) %>%
#   group_by(pathway) %>%
#   summarise(size = length(unique(SNP_ID)),
#             n_genes = length(unique(ensembl_gene_id)))
#   select(pathway, n_genes, size) %>%
#   ungroup() %>%
#   write_csv(., "random_subsets_summary.csv")
# 
# null_summary <- read_csv("random_subsets_summary.csv")

# ## without BCAT genes
# bcat_genes <- read.table("data/processed/pathways/bcat_genes/bcat_genes.txt", header = TRUE)
# 
# feature_sizes_nobcat <- tibble(file = files) %>%
#   separate(file, sep = "/", into = c("dir", "dir2", "dir3", "pathway"), remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE)) %>%
#   unnest(data) %>%
#   mutate(pathway = str_replace(pathway, "null", "c"),
#          pathway = str_replace(pathway, ".txt", "")) %>%
#   group_by(pathway) %>%
#   anti_join(., bcat_genes, by = "SNP_ID") %>%
#   summarise(size_nobcat = length(unique(SNP_ID)),
#             n_genes_nobcat = length(unique(ensembl_gene_id))) %>%
#   select(pathway, n_genes_nobcat, size_nobcat) %>%
#   ungroup() %>%
#   left_join(null_summary, ., by = "pathway") %>%
#   write_csv(., paste0(date, "_random_subsets_summary.csv"))

null_summary <- read_csv("20190212_random_subsets_summary.csv") %>%
  select(pathway, n_genes, size)

kable(head(null_summary))

```

```{r null lr}

### run hashed sections on cluster

# ## list files for null distribution 
# files <- list.files(path = "models/reml_null", pattern = "h2_\\d+.reml$",
#                     full.names = TRUE, recursive = TRUE)
# 
# # calculate likelihood ratio for null distribution
# 
# lr_null <- tibble(file = files) %>%
#   separate(file, sep = "/",
#            into = c("dir", "source", "pathway", "id"), remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE)) %>%
#   unnest(data) %>%
#   filter(Num_Kinships %in% "Alt_Likelihood") %>%
#   mutate(id = str_replace(id, "reml_h2_", ""),
#          id = str_replace(id, ".reml", ""),
#          null_llik = as.numeric(X2)) %>%
#   left_join(., aa_names, by = "id") %>%
#   left_join(., aa_cat, by = "trait") %>%
#   left_join(., gblup_llik, by = c("id", "trait")) %>%
#   select(trait, id, aa_cat, pathway, Num_Kinships, null_llik, gblup_llik) %>%
#   left_join(., null_summary, by = "pathway") %>%
#   mutate(group_size = cut(size, breaks = seq(0, 130000, by = 10000),
#                           include.lowest = TRUE, dig.lab = 10)) %>%
#   mutate(LR = 2 * (null_llik - gblup_llik)) %>%
#   write_csv(., "null_lr.csv")

lr_null <- read_csv("null_lr.csv") %>%
  left_join(., null_summary, by = "pathway") %>%
  select(-n_genes.x, -size.x) %>%
  rename(size = size.y,
         n_genes = n_genes.y)

# add variable for group size (number of SNPs) by intervals of 10000 (for plotting)
lr_null <- lr_null %>%
  mutate(group_size = fct_relevel(group_size, "[0,10000]")) 

head(lr_null)

```

```{r null h2}

# h2_null <- tibble(file = files) %>%
#   separate(file, sep = "/", into = c("dir", "source", "pathway", "id"), remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE, skip = 13)) %>%
#   unnest(data) %>%
#   mutate(id = str_replace(id, "reml_h2_", ""),
#          id = str_extract(id, "\\d+")) %>%
#   left_join(., aa_names, by = "id") %>%
#   select(trait, pathway, Component, Heritability, Her_SD) %>%
#   filter(Component %in% c("Her_K1")) 

h2_null <- read_csv("null_h2.csv")

head(h2_null)

```

```{r combine lr and h2}

# merge LR and H2 results 
null_dist <- left_join(lr_null, h2_null, by = c("trait", "pathway")) %>%
  mutate(Heritability = ifelse(Heritability < 0, 0, Heritability)) %>% # set negative H2 to zero
  # remove observations where h2 > 1
  # remove values where SD of heritability is strange (results in inflated LR)
  filter(Heritability <= 1 & 
           Her_SD > 0 &
           LR >= 0) 

head(null_dist)

```

## LR diagnostics
Expect LR to be distributed as $\chi$^2^~df 1~ based on Wilks' Theorem 

Note that LR values for asp & asp_t do not have asymptotic distributions

```{r null lr diagnostics}

# diagnostic plots

for (i in unique(lr_null$aa_cat)) {
  null_dist %>%
    filter(LR >= 0 & aa_cat == i) %>%
    ggplot(., aes(LR)) + 
    geom_histogram(bins = 50) + 
    facet_wrap( ~ trait, ncol = 4, scales = 'free') 
  
  ggsave(paste0("reports/figures/", date, "_null_hist_", i, ".png"), height = 8, width = 8)
  
  null_dist %>%
    filter(aa_cat == i) %>%
    ggplot(., aes(sample = LR, colour = group_size)) + 
    scale_colour_manual(values = viridis(6)[1:5]) + 
    stat_qq(distribution = qchisq, dparams = list(df = 1)) +
    facet_wrap( ~ trait, scales = "free", ncol = 4) +
    geom_abline(slope = 1, intercept = 0) 
  
  ggsave(paste0("reports/figures/", date, "_null_qqchisq_", i, ".png"), height = 8, width = 8)
  
}

```

### Goodness-of-fit tests (Kolmogorov-Smirnov & Chisq)

```{r test goodness of fit }

# group by trait
# export test statistics

GoF <- null_dist %>% 
  select(trait, LR) %>%
  group_by(trait) %>%
  mutate(
    # return distance (D) for Kolmogorov-Smirnov goodness-of-fit test 
    # df = 1
    D1 = ks.test(LR, pchisq, df = 1)$statistic,
    # df = mixture of 1 and 2 
    D1_2 = ks.test(LR, rchibarsq(length(LR), 2, mix = 0.5))$statistic,
    # df = 2
    D2 = ks.test(LR, pchisq, df = 2)$statistic,
    # return p-value for chi-sq goodness of fit test 
    # df = 1
    X2_1 = suppressWarnings(chisq.test(LR, rchisq(length(LR), 1))$p.value),
    # df = mixture of 1 and 2
    X2_1_2 = suppressWarnings(chisq.test(LR, rchibarsq(length(LR), 2, mix = 0.5))$p.value),
    # df = 2
    X2_2 = suppressWarnings(chisq.test(LR, rchisq(length(LR), 2))$p.value)) %>%
  summarise_all(mean)

GoF <- left_join(GoF, aa_cat, by = "trait") %>%
  select(trait, aa_cat, LR, D1, D1_2, D2, X2_1, X2_1_2, X2_2)

# return table of results 
GoF %>% 
  kable(digits = 4) %>%
  kable_styling(bootstrap_options = "striped")

```

### Plot of Kolmogorov-Smirnov statistic

```{r plot KS statistic}

# plot results
for (i in unique(GoF$aa_cat)) {
  GoF %>%
    select(trait, aa_cat, D1, D1_2, D2) %>%
    filter(aa_cat == i) %>%
    gather(key = df, value = D, -trait, -aa_cat) %>%
    mutate(df = recode(df, D1 = "1", D1_2 = "1.5", D2 = "2")) %>%
    ggplot(., aes(df, D, group = trait)) +
    geom_point() + 
    geom_line() +
    ggtitle(paste(i)) + 
    xlab("degrees of freedom") +
    ylab("Kolmogorov-Smirnov test statistic (D)") + 
    facet_wrap(~ trait, scales = "fixed", ncol = 4) +
    theme_bw() +
    theme(legend.position = "none")
  
  ggsave(paste0("reports/figures/", date, "_KS_stat_", i, ".png"), height = 8, width = 8)
  
}

```

```{r get minimum D value}

GoF %>%
  select(trait, aa_cat, D1, D1_2, D2) %>%
  gather(key = df, value = D, -trait, -aa_cat) %>%
  group_by(trait) %>%
  mutate(df = recode(df, D1 = "1", D1_2 = "1.5", D2 = "2")) %>%
  slice(which.min(D)) %>%
  ungroup() %>%
  write_csv(paste0("reports/", date, "_goodness_of_fit.csv"))

```

## 95 quantile for LR 

```{r 95 quantile for LR}
## 95% LR threshold for plotting
lr_fit <- NULL
coefs <- NULL

for (i in unique(null_dist$trait)) {
  lr_fit[[i]] <- rqss(as.numeric(LR) ~ qss(size, constraint = "I"), 
                      data = null_dist[null_dist$trait==i,], tau = 0.95, lambda = 500)
  coefs[[i]] <- data.frame(lr_fit[[i]]$qss$size$xyz) %>%
    mutate(lr_95 = X2 + lr_fit[[i]]$coef[1],
           size = X1) %>%
    select(size, lr_95)
}

lr_threshold <- rbindlist(coefs, idcol = "trait")

null_dist <- null_dist %>%
  left_join(., lr_threshold, by = c("trait", "size")) %>%
  mutate(col = LR >= lr_95)

```

## 95 quantile for heritability 

```{r 95 quantile for H2}
h2_fit <- NULL
h2_coefs <- NULL

for (i in unique(null_dist$trait)) {
  h2_fit[[i]] <- rqss(as.numeric(Heritability) ~ qss(size, constraint = "I"), 
                      data = null_dist[null_dist$trait==i,], tau = 0.95, lambda = 500)
  h2_coefs[[i]] <- data.frame(h2_fit[[i]]$qss$size$xyz) %>%
    mutate(h2_95 = X2 + h2_fit[[i]]$coef[1],
           size = X1) %>%
    select(size, h2_95)
}

h2_threshold <- rbindlist(h2_coefs, idcol = "trait")

```

## Plot null distribution with 95 quantile for H2

```{r h2 95 threshold}

for (i in unique(null_dist$aa_cat)) {
  null_dist %>%
    filter(aa_cat == i) %>%
    ggplot(., aes(size, Heritability, colour = col)) +
    scale_colour_manual(values = c("#a6bddb", "#1c9099"),
                        labels = c(expression(paste(LR < LR[95])), 
                                   expression(paste(LR > LR[95])))) + 
    geom_point(position = position_dodge(width = 0.3), aes(size = col), alpha = 0.75) + 
    scale_size_manual(guide = "none", values = c(0.1, 0.5)) + 
    geom_quantile(method = "rqss", lambda = 500, quantiles = c(0.95),
                  formula = as.formula(y ~ qss(x, constraint = "I")),
                  aes(linetype = factor(..quantile..)),
                  colour = "black", show.legend = FALSE) + 
    geom_abline(aes(intercept = 0, slope = 1/213612, linetype = "dashed")) + 
    scale_linetype_manual(labels = c(expression(paste(H[95]^2)), 
                                     "infinitesimal"),
                          values = c(1,2)) +
    facet_wrap( ~ trait, ncol = 4) +
    xlim(0, 50000) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.title = element_blank(),
          legend.text.align = 0)
  
     ggsave(paste0("reports/figures/", date, "_null_95_", i, ".png"), height = 8, width = 7)
}

```

## Multiblup thresholds for significance
Use rqss.predict to determine what the 95% cut-off is for each pathway

```{r 95 threshold for multiblup}

# files <- list.files(path = "data/processed/pathways", pattern = "*.txt", full.names = TRUE,
#                     recursive = TRUE)
# 
# pathways <- tibble(file = files) %>%
#   separate(file, sep = "/", into = c("dir", "dir2", "dir3", "pathway", "pathway2"), 
#            remove = FALSE) %>%
#   mutate(data = lapply(file, read.table, header = TRUE),
#          data = lapply(data, mutate_all, as.character)) %>%
#   unnest(data) %>%
#   group_by(pathway) %>%
#   summarise(size = length(unique(SNP_ID)),
#             n_genes = length(unique(ensembl_gene_id.x))) %>%
#   select(pathway, n_genes, size) %>%
#   ungroup() %>%
#   write_csv(., paste0("reports/", date, "_pathways_summary.csv"))

pathways <- read_csv("reports/20190212_pathways_summary.csv") %>%
  filter(!pathway == "glycolysis")

multiblup_thresholds <- pathways %>%
  group_by(pathway) %>%
  mutate(stat = list(c(h2_95 = as.data.frame(lapply(h2_fit[1:49], predict, 
                                                     data.frame(size = size))),
                        lr_95 = as.data.frame(lapply(lr_fit[1:49], predict, 
                                                     data.frame(size = size)))))
         ) %>%
  mutate(stat = map(stat, bind_rows)) %>%
  unnest() %>%
  gather(., id, value, -pathway, -n_genes, -size) %>%
  separate(col = id, into = c("stat", "trait"), sep = "\\.") %>%
  spread(., stat, value)

head(multiblup_thresholds)
  
write_csv(multiblup_thresholds, paste0("reports/", date, "_multiblup_thresholds.csv"))

```

